### PR TITLE
Add support for file type #34

### PR DESCRIPTION
**Pull request by zbarbuto**
Also closes https://github.com/strongloop/loopback-component-explorer/issues/103

### Description

Allows `File` type to be turned into a file upload field.

#### Related issues

- #34
- https://github.com/strongloop/loopback-component-explorer/issues/103

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by bajtos** [comment]
Hello @zbarbuto, thank you for the pull request! The changes looks mostly good to me, however I think we need to be more careful about the new type name.

TL;DR: let's use `file` as the type name please.

Background: see the discussion in https://github.com/strongloop/loopback/issues/2554. As of https://github.com/strongloop/strong-remoting/pull/370, only `file` is recognized as a type representing a file (arbitrary stream). The name `File` is left available for an application-defined model.

**Comment by zbarbuto** [comment]
@bajtos Fair enough. Updated.


>The name File is left available for an application-defined model.

Interesting. I thought this wasn't the case as the model name gets lower-cased and throws an error if it is either 'file' or 'File':
https://github.com/strongloop/strong-remoting/commit/2c4cc6eae4db68fb0b30df91966a92e3413377e8

File: lib/codegen/spec-converter.js [hunk 1]
```diff
@@ -141,11 +141,6 @@ function convertApi(v2, apiDeclaration, definitions) {
           } else {
             _.extend(target, props);
           }
-          // Files hasn't made it over yet.
-          //
-          if (target.type === 'File') {
```
Comment (zbarbuto) [inline lib/codegen/spec-converter.js]: @bajtos Also removed this bock as doesn't seem to be the case anymore after this PR?

Comment (bajtos) [inline lib/codegen/spec-converter.js]: Yeah, makes sense. Just note that AFAIK, strong-remoting does not support file uploads (input parameters of `file` type) yet - see https://github.com/strongloop/loopback/issues/3266

File: lib/specgen/type-registry.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ function TypeRegistry() {
     },
   });
   this.registerLoopbackType('DateString', {type: 'string', format: 'date-time'});
+  this.registerLoopbackType('File', {type: 'file'});
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: I think the type name should be `file` for consistency with other places where we are using `file` too.

File: test/specgen/route-helper.test.js [hunk 3]
```diff
@@ -103,6 +103,20 @@ describe('route-helper', function() {
       });
   });
 
+  it('converts { type: File\' } to { schema: { type: \'file\' } }', function() {
```
Comment (bajtos) [inline test/specgen/route-helper.test.js]: ditto, `type: file`

File: test/specgen/route-helper.test.js [hunk 4]
```diff
@@ -103,6 +103,20 @@ describe('route-helper', function() {
...
+    var TestModel = loopback.createModel('TestModel', {street: String});
+    var entry = createAPIDoc({
+      accepts: [
+        {name: 'changes', type: 'File'},
```
Comment (bajtos) [inline test/specgen/route-helper.test.js]: and here too

**Comment by bajtos** [comment]
> > The name File is left available for an application-defined model.

> Interesting. I thought this wasn't the case as the model name gets lower-cased loopback throws an error if the application tries to define a model called 'file' or 'File':
strongloop/strong-remoting@2c4cc6e

> Or by 'application-defined' do you mean core to loopback?

By application-defined I mean a model defined by a 3rd-party/user application using LoopBack as the framework. E.g. https://github.com/strongloop/loopback-sandbox (it does not have any File model defined though).

After I looked around in strong-remoting, I am little bit confused about the benefits of your proposed changes TBH.  If strong-remoting which treats `file` as a built-in/reserved type that's not supported for input arguments yet, then what's the point of supporting input parameters with `file` type in swagger spec? What am I missing?

**Comment by zbarbuto** [comment]
>After I looked around in strong-remoting, I am little bit confused about the benefits of your proposed changes TBH. If strong-remoting which treats file as a built-in/reserved type that's not supported for input arguments yet, then what's the point of supporting input parameters with file type in swagger spec? What am I missing?

The crux of it is: I want to be able to use the api explorer in conjunction with https://github.com/strongloop/loopback-component-storage in order to test file upload endpoints. 

At the moment, I'm having to use third-party API tools just to test my file upload endpoints. Since swagger definitely [supports file uploads](https://github.com/strongloop/loopback-swagger/blob/2681244e8b17dabb52702a884c7d38fb611fa780/example/specs/1.2/pet.json#L304) I thought it would just be a case of modifying the swagger gen to actually output a file upload type in the event of an `accepts: type: file` situation, then allow swagger to take it from there.

**Comment by bajtos** [comment]
@zbarbuto first of all, sorry for the delay.

> The crux of it is: I want to be able to use the api explorer in conjunction with https://github.com/strongloop/loopback-component-storage in order to test file upload endpoints.

That makes sense 👍 

However, I think the fix will require more changes than just here in loopback-swagger.

Take a look at how the file upload method is defined in [lib/storage-service.js](https://github.com/strongloop/loopback-component-storage/blob/4572305831d5020d3f6b2410fcb5fdb79172fe2c/lib/storage-service.js#L339-L346):

```js

StorageService.prototype.upload.shared = true;
StorageService.prototype.upload.accepts = [
  {arg: 'req', type: 'object', 'http': {source: 'req'}},
  {arg: 'res', type: 'object', 'http': {source: 'res'}},
];
StorageService.prototype.upload.returns = {arg: 'result', type: 'object'};
StorageService.prototype.upload.http =
{verb: 'post', path: '/:container/upload'};
```

It is accepting a full `request` object, there is no mention of file uploads. As a result, strong-remoting and loopback-swagger don't have any visibility into what kind of data is the method expecting, we cannot tell that this method expect file uploads.

The proper solution is to implement support for file uploads in strong-remoting and then rework loopback-component-storage to let strong-remoting handle file uploads, instead of the current implementation where loopback-component-storage is parsing multipart requests itself (see [lib/storage-handler.js](https://github.com/strongloop/loopback-component-storage/blob/4572305831d5020d3f6b2410fcb5fdb79172fe2c/lib/storage-handler.js#L28-L112)). With that two changes in place, and this pull request landed, you should finally be able to upload files directly from our swagger-ui powered Explorer.

Thoughts?

**Comment by zbarbuto** [comment]
@bajtos you're correct in that it would take multiple changes to get it to work with storage component out of the box. However in my case I'm not using the default methods provided by storage component. I've wrapped them in a custom endpoint. So what this PR achieves is allowing arbitrary endpoints to specify a file type for swagger. Then a separate issue/PR could be made for storage to support this by default.

**Comment by bajtos** [comment]
> So what this PR achieves is allowing arbitrary endpoints to specify a file type for swagger. Then a separate issue/PR could be made for storage to support this by default.

Sure, makes sense.

**Comment by bajtos** [comment]
I have rebased you patch on top of the latest master and squashed the commits into a single one.

Landed, thank you for the contribution!

**Comment by bajtos** [comment]
For posterity, https://github.com/strongloop/loopback/issues/3266 is keeping track of the feature allowing remote method to accept file uploads.
